backend/src/main/scala/cromwell/backend/callcaching/CallCachingActor.scala

```diff
@@ -0,0 +1,202 @@
+package cromwell.backend.callcaching
+
+import akka.actor.{LoggingFSM, Props}
+import cromwell.backend.BackendJobDescriptor
+import cromwell.backend.BackendJobExecutionActor.{BackendJobExecutionResponse, SucceededResponse}
+import cromwell.backend.callcaching.CallCachingActor._
+import cromwell.core._
+import wdl4s.RuntimeAttributes
+
+import scala.concurrent.ExecutionContext
+
+/**
+  * The call caching actor is a total maverick!
+  *
+  * It goes off on its own as soon as it is created and decides whether the JobDescriptor it has received is a cache hit.
+  */
+case class CallCachingActor
+(
+  backendJobDescriptor: BackendJobDescriptor,
+  hashFileFunction: String => String,
+  hashRuntimeAttributesFunction: RuntimeAttributes => String,
+  readFromCache: Boolean,
+  writeToCache: Boolean
+) extends LoggingFSM[CCAState, CCAData] {
+
+  implicit val ec: ExecutionContext = context.dispatcher
+
+  private val anyWorkRequired = readFromCache || writeToCache
+
+  if (anyWorkRequired) calculateHash()
+
+  startWith(
+    if (anyWorkRequired) DeterminingCacheHit else NeverDidAnything,
+    NoData
+  )
+
+  when(DeterminingCacheHit) {
+    case Event(CheckCache, _) =>
+      sender ! CallCacheCheckInProgress
+      stay()
+    case Event(CacheHitInternal(cacheResultId), _) if readFromCache =>
+      readEntryFromDatabase(cacheResultId)
+      goto(FetchingJobResult)
+    case Event(CacheHitInternal(_), _) if !readFromCache =>
+      goto(UnreadCacheHit)
+    case Event(CacheMissInternal(hashes), _) if writeToCache =>
+      goto(AwaitingJobResultAndAllHashes) using WithHashes(hashes)
```

Comment:
Not clear why this sets the state data to `WithHashes(hashes)`?  It doesn't look like this is considered in `AwaitingJobResultAndAllHashes`.